### PR TITLE
Refactor UDF management process to support table model user-defined function

### PR DESCRIPTION
**Pull request by Cpaulyz**
## Description

1. Refactor UDF management process. Most important change is to introduce an enum type `Model`. UDFTable use `Pair<Model, String>` as hash key to support table model user-defined function now. 
2. Introduce pre create and pre drop process to avoid some corner cases  due to partial create/drop. Function with **UNAVAILABLE** state in `SHOW FUNCTIONS` will perform some unexpected behaviors.

See https://timechor.feishu.cn/docx/TCyYdm8pYowe0HxEGb9c2UfenCf for more details.

### CODE REVIEW

**Comment by OneSizeFitsQuorum** [comment]
What a spectacular work 🦾! Fabulous design and implementation⚡️⚡️🔥

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFTable.java [hunk 1]
```diff
@@ -19,99 +19,101 @@
...
 public class UDFTable {
-  private final Map<String, UDFInformation> udfInformationMap;
 
-  /** maintain a map for creating instance */
-  private final Map<String, Class<?>> functionToClassMap;
+  /** functionName -> information * */
+  private final Map<Pair<Model, String>, UDFInformation> udfInformationMap;
+
+  /** maintain a map for creating instance, functionName -> class */
+  private final Map<Pair<Model, String>, Class<?>> functionToClassMap;
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFTable.java:48]: Why not using Map<Model, Map<String, UDFInformation>>? In this way, you won't need to iterate all map entry while trying to all functions of one kind of model.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public enum UDFType {
+  TREE_EXTERNAL((byte) 0),
+  TREE_BUILT_IN((byte) 1),
+  TREE_UNAVAILABLE((byte) 2),
+  TABLE_EXTERNAL((byte) 3),
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java:32]:
```suggestion
  TABLE_AVAILABLE((byte) 3),
```

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+package org.apache.iotdb.commons.udf;
+
+import org.apache.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+public enum UDFType {
+  TREE_EXTERNAL((byte) 0),
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java:29]:
```suggestion
  TREE_AVAILABLE((byte) 0),
```

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  TREE_BUILT_IN((byte) 1),
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFType.java:30]: add some comments for this enum to explain that this won't appear in previous snapshot file or raft log, just a place holder for some unforeseen circumstances.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/service/UDFExecutableManager.java [hunk 5]
```diff
@@ -53,6 +58,47 @@ public static synchronized UDFExecutableManager setupAndGetInstance(
     return INSTANCE;
   }
 
+  /** check whether local jar is correct according to md5 */
+  public boolean isLocalJarConflicted(UDFInformation udfInformation) throws UDFManagementException {
+    String functionName = udfInformation.getFunctionName();
+    // A jar with the same name exists, we need to check md5
+    String existedMd5 = "";
+    String md5FilePath = functionName + ".txt";
```
Comment (JackieTien97) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/service/UDFExecutableManager.java:66]: `.txt` already used previously? if not, I think `.md5` suffix may be better

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/metadata/ShowFunctionsTask.java [hunk 6]
```diff
@@ -55,9 +61,29 @@
 
 public class ShowFunctionsTask implements IConfigTask {
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/metadata/ShowFunctionsTask.java:62]: will this class be used for table model? it seems that you don't deal with that

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/service/UDFExecutableManager.java (hunk 5 above)
Comment (Cpaulyz) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/service/UDFExecutableManager.java:66]: > `.txt` already used previously? if not, I think `.md5` suffix may be better

Yes. The code is copied from [UDFManagementService.java](https://github.com/apache/iotdb/pull/14135/files/656dc4c48bf6c3acd20f215729f7616aca636c0f#diff-f9da4fa77c651fa972a95aa9ae795f2dfb96a795d5cdd7e0fd5022be23537f0e)

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/metadata/ShowFunctionsTask.java (hunk 6 above)
Comment (Cpaulyz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/metadata/ShowFunctionsTask.java:62]: CREATE/DROP/SHOW FUNCTION statement of table model will be supported in the next PR. This PR just want to refactor common codes about UDF management and make sure it won't affect tree model.

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/UDFManager.java [hunk 7]
```diff
@@ -83,35 +86,43 @@ public TSStatus createFunction(TCreateFunctionReq req) {
...
-      if (dataNodesStatus.getCode() != TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
-        return dataNodesStatus;
-      }
+      final boolean needToSaveJar = isUsingURI && udfInfo.needToSaveJar(jarName);
 
+      LOGGER.info("Start to add UDF [{}] in UDF_Table on Config Nodes", udfName);
       CreateFunctionPlan createFunctionPlan =
           new CreateFunctionPlan(udfInformation, needToSaveJar ? new Binary(jarFile) : null);
+
+      udfInformation.setAvailable(true);
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/UDFManager.java:102]: It should be true here. We should set it to true after executing `CreateFunctionPlan`(pre-creation) with avaliable=false.

[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/manager/UDFManager.java:102]: It should be false. Fixed.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFTable.java (hunk 1 above)
Comment (Cpaulyz) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/UDFTable.java:48]: Done. Use `Map<Model, Map<String, UDFInformation>>` now

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/function/DropFunctionPlan.java [hunk 8]
```diff
@@ -31,29 +32,37 @@
...
   @Override
   protected void serializeImpl(DataOutputStream stream) throws IOException {
     stream.writeShort(getType().getPlanType());
+    ReadWriteIOUtils.write(model.getValue(), stream);
     ReadWriteIOUtils.write(functionName, stream);
   }
 
   @Override
   protected void deserializeImpl(ByteBuffer buffer) throws IOException {
+    model = Model.findByValue(ReadWriteIOUtils.readInt(buffer));
```
[the remark this answers is not in the document]
Comment (Cpaulyz) [reply iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/consensus/request/write/function/DropFunctionPlan.java]: Fixed. Split it into DropTableModelFunctionPlan and DropTreeModelFunctionPlan

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**